### PR TITLE
Only throw instances of dart:core.Error

### LINKED ISSUE #239: Flag throwing non Error/Exception types
```dart
throw 'some string';
```

or `null`, number, etc.

### PR DESCRIPTION
**Pull request by alexeieleusis**
fixes #239

### CODE REVIEW

File: lib/src/rules/only_throw_error.dart [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  @override
+  void visitThrowExpression(ThrowExpression node) {
+    if (node.expression is Literal) {
+      rule.reportLint(node);
+      return;
+    }
+
+    if (!DartTypeUtilities.extendsClass(
+        node.expression.bestType, 'Error', 'dart.core')) {
```
Comment (bwilkerson) [inline lib/src/rules/only_throw_error.dart]: We need to also allow throwing instances of subclasses of Exception. Then we should try running this rule over some large code bases to see whether there are any false positives.

File: lib/src/util/dart_type_utilities.dart [hunk 2]
```diff
@@ -40,6 +40,12 @@ class DartTypeUtilities {
         element.allSupertypes.any(predicate);
   }
 
+  static bool extendsClass(DartType type, String className, String library) =>
```
Comment (bwilkerson) [inline lib/src/util/dart_type_utilities.dart]: I suspect that we want to allow the `type` to either extend or implement the given class, which means that this has to handle mixins and interfaces as well as superclasses.

File: test/integration_test.dart [hunk 3]
```diff
@@ -238,6 +238,38 @@ defineTests() {
       });
     });
 
+    solo_group('only_throw_error', () {
```
Comment (bwilkerson) [inline test/integration_test.dart]: The ''solo' prefix, both here and on the test, need to be removed.

File: lib/src/util/dart_type_utilities.dart (hunk 2 above)
Comment (alexeieleusis) [inline lib/src/util/dart_type_utilities.dart]: fixed and included `dynamic`

File: lib/src/rules/only_throw_error.dart (hunk 1 above)
Comment (alexeieleusis) [inline lib/src/rules/only_throw_error.dart]: fixed, the only "false" positive I found on the SDK codebase is `throw new DomException.jsInterop`  but when I tried opening the file it didn't exist.

File: test/integration_test.dart (hunk 3 above)
Comment (alexeieleusis) [inline test/integration_test.dart]: ups, fixed

File: lib/src/util/dart_type_utilities.dart [hunk 4]
```diff
@@ -40,6 +40,12 @@ class DartTypeUtilities {
...
+      type != null && type.name == className &&
+          type.element.library.name == library ||
+          (type is InterfaceType &&
+              extendsClass(type.superclass, className, library));
```
Comment (bwilkerson) [inline lib/src/util/dart_type_utilities.dart]: Admittedly, it's a strange corner case, but should we support using Error and Exception as mixins? For example:

```
class E extends Object with Exception {
  static throw() {
    throw new E();
  }
}
```

File: lib/src/rules/only_throw_error.dart [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+  }
+}
+
+const _library = 'dart.core';
+const _errorClassName = 'Error';
+const _exceptionClassName = 'Exception';
+
+bool _isThrowable(DartType type) {
+  return type.isDynamic ||
```
Comment (bwilkerson) [inline lib/src/rules/only_throw_error.dart]: Perhaps add a comment noting that the performance here could be better if we searched the type hierarchy once and quit once we've found any of a number of classes.

**Comment by bwilkerson** [comment]
Other than minor nits, lgtm.

File: lib/src/util/dart_type_utilities.dart (hunk 4 above)
Comment (alexeieleusis) [inline lib/src/util/dart_type_utilities.dart]: This was already a handled case, added a test :)

File: lib/src/rules/only_throw_error.dart (hunk 5 above)
Comment (alexeieleusis) [inline lib/src/rules/only_throw_error.dart]: Changed implementation, I also thought about this, but I was not sure.

**Comment by bwilkerson** [comment]
lgtm

**Comment by alexeieleusis** [comment]
Thank you!

**Pull request merged**